src/libcrun/linux.c

```diff
@@ -1058,6 +1059,63 @@ has_mount_for (libcrun_container_t *container, const char *destination)
   return false;
 }
 
+/* Get or create the global shared empty directory path for masked paths optimization. */
+static int
+get_shared_empty_dir_path (libcrun_container_t *container, char **empty_dir_path, libcrun_error_t *err)
+{
+  cleanup_free char *run_dir = NULL;
+  int ret;
+
+  ret = get_run_directory (&run_dir, container->context->state_root, err);
+  if (UNLIKELY (ret < 0))
+    return ret;
+
+  /* Create path to global empty directory */
+  ret = append_paths (empty_dir_path, err, run_dir, ".empty-directory", NULL);
+  if (UNLIKELY (ret < 0))
+    return ret;
+
+  /* Ensure the empty directory exists */
+  ret = crun_ensure_directory (*empty_dir_path, 0555, false, err);
+  if (UNLIKELY (ret < 0))
+    return ret;
+
+  return 0;
+}
+
+static int
+mount_masked_dir (libcrun_container_t *container, int pathfd, const char *rel_path, libcrun_error_t *err)
+{
+  cleanup_free char *empty_dir_path = NULL;
+  libcrun_error_t tmp_err = NULL;
+  int ret;
+
+  ret = get_shared_empty_dir_path (container, &empty_dir_path, &tmp_err);
+  if (ret >= 0)
+    {
+      ret = do_mount (container, empty_dir_path, pathfd, rel_path, NULL, MS_BIND | MS_RDONLY, NULL, LABEL_MOUNT, &tmp_err);
+      if (LIKELY (ret >= 0))
+        return ret;
+
+      if (tmp_err != NULL)
```

Comment:
Maybe the if statement


```
  if (tmp_err != NULL)
```

 is unnecessary? 

I created an issue related to this

* https://github.com/containers/crun/issues/1860